### PR TITLE
[WIP] go1.15 support

### PR DESCRIPTION
**Pull request by justaugustus**
Signed-off-by: Stephen Augustus <saugustus@vmware.com>

/hold testing go1.15beta1 release for https://github.com/kubernetes/kubernetes/pull/93264
cc: @kubernetes/release-engineering

### CODE REVIEW

File: repos.bzl [hunk 1]
```diff
@@ -29,6 +34,57 @@ def configure(minimum_bazel_version = None, rbe_name = "rbe_default", go_version
     protobuf_deps()  # No options
 
     go_rules_dependencies()  # No options
+
+    go_download_sdk(
```
Comment (fejta) [inline repos.bzl:38]: Please make this optional as I believe it forces everyone to use 1.15beta1.

For example:

```python
EXTRA_VERSIONS = {"go1.15beta1": {
  "darwin_amd64": ("go1.15beta1.darwin-amd64.tar.gz", "4ee49...bd7e"),
  "freebsd_386": ("go1.15beta1.freebsd-386.tar.gz", "77bc...5c94"),
  ...
}}

def configure(... extra_version=None):
  ...
  if extra_version:   
      go_download_sdk(name="go_sdk", sdks=EXTRA_VERSIONS[extra_version])
      go_register_toolchains(go_version = extra_version, nogo = nogo)
  else 
      go_register_toolchains(go_version = go_version, nogo = nogo)
```

File: go.mod [hunk 2]
```diff
@@ -3,9 +3,9 @@ module k8s.io/repo-infra
...
 require (
-	github.com/bazelbuild/bazel-gazelle v0.20.0
+	github.com/bazelbuild/bazel-gazelle v0.21.1
 	github.com/bazelbuild/buildtools v0.0.0-20200228172928-c9d9e342afdb
-	github.com/golangci/golangci-lint v1.25.0
-	golang.org/x/build v0.0.0-20200302185339-bb8466fe872a
-	k8s.io/klog/v2 v2.0.0
+	github.com/golangci/golangci-lint v1.29.0
+	golang.org/x/build v0.0.0-20200720211405-d0191c8228ec
+	k8s.io/klog/v2 v2.3.0
```
Comment (fejta) [inline go.mod:10]: Can we move all these version bumps to their own PR?

File: repos.bzl (hunk 1 above)
Comment (justaugustus) [reply repos.bzl:38]: I like this! Will do.

File: go.mod (hunk 2 above)
Comment (justaugustus) [reply go.mod:10]: Yep! This was more so yak shaving in support of the [k/k go1.15 bump](https://github.com/kubernetes/kubernetes/pull/93264).

I'll split it out.

Comment (justaugustus) [reply go.mod:10]: Deps bump in https://github.com/kubernetes/repo-infra/pull/198.

File: repos.bzl (hunk 1 above)
Comment (justaugustus) [reply repos.bzl]: Fixed up!

**Comment by justaugustus** [comment]
Hey @jayconrod, I'm trying out your fix in https://github.com/kubernetes/repo-infra/commit/56161307d45d6ab30dc17983ba653b9ad9b8022d (https://github.com/bazelbuild/rules_go/pull/2579), but getting some errors.

Have you seen this before: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/repo-infra/197/pull-repo-infra-tests/1286073407053500416

Is it maybe related to the fact that we're using the rules_go 0.22.z patch series instead of 0.23.z?

**Comment by justaugustus** [comment]
c695b31a4320606f1c1a2b0d1fdb41e728bd45f4 pulls in https://github.com/bazelbuild/rules_go/pull/2579.

I'm going to keep this PR roughly in this state and tackle the [alternate go versions support](https://github.com/kubernetes/repo-infra/pull/197#discussion_r458293039) in a separate PR (since that was causing https://github.com/kubernetes/kubernetes/pull/93264 to fail).

[Example](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93264/pull-kubernetes-bazel-test/1285737079153627136):

```
 DEBUG: Rule 'io_k8s_repo_infra' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1595370464 -0400"
DEBUG: Call stack for the definition of repository 'io_k8s_repo_infra' which is a git_repository (rule definition at /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/external/bazel_tools/tools/build_defs/repo/git.bzl:195:18):
 - <builtin>
 - /home/prow/go/src/github.com/kubernetes/kubernetes/WORKSPACE:20:1
Loading: 0 packages loaded
Loading: 0 packages loaded
ERROR: /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/external/io_bazel_rules_go/go/private/sdk.bzl:251:9: Traceback (most recent call last):
	File "/home/prow/go/src/github.com/kubernetes/kubernetes/WORKSPACE", line 36
		repo_infra_configure(<2 more arguments>)
	File "/bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/external/io_k8s_repo_infra/repos.bzl", line 91, in repo_infra_configure
		go_register_toolchains(<2 more arguments>)
	File "/bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/external/io_bazel_rules_go/go/private/sdk.bzl", line 251, in go_register_toolchains
		fail(<1 more arguments>)
go_version set after go sdk rule declared (go_sdk)
ERROR: error loading package '': Encountered error while reading extension file 'repositories/repositories.bzl': no such package '@io_bazel_rules_docker//repositories': error loading package 'external': Could not load //external package
ERROR: error loading package '': Encountered error while reading extension file 'repositories/repositories.bzl': no such package '@io_bazel_rules_docker//repositories': error loading package 'external': Could not load //external package 
```

**Comment by jayconrod** [comment]
> Is it maybe related to the fact that we're using the rules_go 0.22.z patch series instead of 0.23.z?

@justaugustus Yes, that looks like the problem. In `@io_k8s.repo_infra//:repos.bzl`, a the repository `org_golang_x_tools` is declared at a specific version, and the patch `@io_bazel_rules_go//third_party:org_golang_x_tools.patch` is applied. That patch changes with the newer version of rules_go, and it no longer applies cleanly to this version.

I'll backport this fix and tag a 0.22.x point release today, so hopefully you won't need to update anything else.

**Comment by justaugustus** [comment]
(Rebased to pull in rules_go@v0.22.9 --> https://github.com/kubernetes/repo-infra/pull/201)

**Comment by justaugustus** [comment]
Superseded by https://github.com/kubernetes/repo-infra/pull/202.
go1.15beta1 support available in https://github.com/kubernetes/repo-infra/releases/tag/v0.0.9 (via `override_go_version`)
/close
